// bookReadAnimesh/DynamoInterface.js
/**
 * Created by ajain13 on 6/20/2017.
 */
/**
 * Interfaces with AWS- Dynamo Db for storing skill specific data, session etc.
 */

var AWS = require('aws-sdk');
var docClient = new AWS.DynamoDB.DocumentClient();
AWS.config.update(
    {
        region:"us-east-1",
    }
);

module.exports.read=function readSessionID(userId, callback) {
    AWS.config.update(
        {
            'region':'us-east-1',
        }
    );
    var params = {
        TableName: "aimreader",
        Key: {
            "userid": userId
        }
    };
    docClient.get(params, function (err, data) {
        if (err) {
            callback(false, err);
            console.log("Unable to read item. JSON:", JSON.stringify(err, null, 2));

        }
        else {
            console.log(data);
            if (Object.keys(data).length == 0) {
                callback(false);
            }
            else {
                callback(data);
            }

        }
    });
}

module.exports.deleteRecord=function deleteRecord(userId) {
    AWS.config.update(
        {
            'region':'us-east-1',
        }
    );
    var params = {
        TableName: "aimreader",
        Key: {
            "userid": userId
        }
    };
    docClient.delete(params, function(err, data) {
        if (err) {
            console.error("Unable to delete item. Error JSON:", JSON.stringify(err, null, 2));
        } else {
            console.log("Delete Item succeeded:", JSON.stringify(data, null, 2));
        }
    });
}
module.exports.create=function createSession(userId,value, callback) {
    var params = {
        TableName: "aimreader",
        Item: {
            "userid": userId,
            "session": value
        }
    };
    docClient.put(params, function (err, data) {
        if (err) {
            console.log("Error create session entry: " + err);
        }
        else {
            console.log("Added Item:", JSON.stringify(data, null, 2));
            callback(true);
        }
    });
}
